clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java

```diff
@@ -0,0 +1,198 @@
+import org.apache.logging.log4j.Logger;
+
+
+public class HeartbeatMonitoringService extends AbstractVeniceService {
+  private final Thread reportingThread;
+  private static final Logger LOGGER = LogManager.getLogger(HeartbeatMonitoringService.class);
+  public static final int DEFAULT_REPORTER_THREAD_SLEEP_INTERVAL_SECONDS = 60;
+
+  // This default value SHOULD be a number that will trip an alert. Alerting on this should be duration based,
+  // so starting off with a large value is fine that is cleared once we get the next heartbeat
+  public long DEFAULT_SENTINEL_HEARTBEAT_TIMESTAMP = System.currentTimeMillis() - 21600000; // 6 hours ago
+
+  private final Set<String> regionNames;
+  private final String localRegionName;
+
+  // store -> version -> partition -> region -> timestamp
+  private final Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> followerHeartbeatTimeStamps;
+  private final Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> leaderHeartbeatTimeStamps;
+  HeartbeatVersionedStats versionStatsReporter;
+
+  public HeartbeatMonitoringService(
+      MetricsRepository metricsRepository,
+      ReadOnlyStoreRepository metadataRepository,
+      Set<String> regionNames,
+      String localRegionName) {
+    this.regionNames = regionNames;
+    this.localRegionName = localRegionName;
+    this.reportingThread = new HeartbeatReporterThread();
+    followerHeartbeatTimeStamps = new VeniceConcurrentHashMap<>();
+    leaderHeartbeatTimeStamps = new VeniceConcurrentHashMap<>();
+    versionStatsReporter = new HeartbeatVersionedStats(
+        metricsRepository,
+        metadataRepository,
+        () -> new HeartbeatStat(new MetricConfig(), regionNames),
+        (aMetricsRepository, s) -> new HeartbeatStatReporter(aMetricsRepository, s, regionNames));
+  }
+
+  private synchronized void initializeEntry(
+      Map<String, Map<Integer, Map<Integer, Map<String, Long>>>> heartbeatTimestamps,
+      Version version,
+      int partition) {
+    // We don't monitor heartbeat lag for non hybrid versions
+    if (version.getHybridStoreConfig() == null) {
+      return;
+    }
+    if (heartbeatTimestamps.get(version.getStoreName()) == null) {
+      heartbeatTimestamps.put(version.getStoreName(), new VeniceConcurrentHashMap<>());
+    }
+    if (heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()) == null) {
+      heartbeatTimestamps.get(version.getStoreName()).put(version.getNumber(), new VeniceConcurrentHashMap<>());
+    }
+    if (heartbeatTimestamps.get(version.getStoreName()).get(version.getNumber()).get(partition) == null) {
+      Map<String, Long> regionTimestamps = new VeniceConcurrentHashMap<>();
+      if (version.isActiveActiveReplicationEnabled()) {
+        for (String region: regionNames) {
+          regionTimestamps.put(region, DEFAULT_SENTINEL_HEARTBEAT_TIMESTAMP);
+        }
+      } else {
+        regionTimestamps.put(localRegionName, DEFAULT_SENTINEL_HEARTBEAT_TIMESTAMP);
```

Comment:
Essentially this means if no heartbeat is received between `initializeEntry` and `record` (up to 60s with current configs) the lag metric will get a spike of 6h. This could generate a lot of spikes/noise when ingestion is just started or rebalanced. We could set up monitoring accordingly to fire only when observing a lag that exceeds threshold for more than let's say 5min. However, I'm wondering if we could just initialize it with `-1` and have the monitoring emit -1 to indicate no heartbeat have been received yet. Since we don't really about the lag without any heartbeat. `DEFAULT_SENTINEL_HEARTBEAT_TIMESTAMP` is just an arbitrary value and could lead to confusion/miscommunication during an incident where oncall will see a lag of 6h+delta but in reality the exact lag could be anything and needs to be manually verified.

Reply:
So the problem with -1 is that our alerting is threshhold and duration based.  I'm not sure if we have a good way to set up "this number should be between 0 and 20k" in a single alert.

Duration is what we would leverage.  We would spike up to 6 hours and increase until we get the next heartbeat.  Something like "lag should not be above 1 hour for 3 hours" or however long we think it should take for ingestion to consume even one event or catch up.

Reply:
Oh!  Also it's a max.  If we use -1, then it'll get screened out by healthy ingesting partitions.

Reply:
I see... was just trying to see if there's a way to avoid these spikes if we have sufficient signals to confirm they are false positive.

Reply:
After musing on it, I think theres probably three approaches we could try.  They are:

- emit -1, but augment Max to give special accordance to the value of -1.  This would mean that it doesn't matter what other big value is out there, if -1 is in there we report it.  A conditional MAX I guess
- Emit a seperate metric for those partitions where we haven't gotten a heartbeat for a partition that we expect it for.
- Commit heartbeats to disk and emit based on the disk value, thus keeping the lag between restarts.  Defaulting to a large number when theres nothing on disk.

Reply:
I guess for V1 we can leave it as it is. If we find it inconvenient or noisy can always improve it later.